src/databricks/labs/ucx/source_code/redash.py

```diff
@@ -1,24 +1,129 @@
+import logging
+from collections.abc import Iterator
+
 from databricks.sdk import WorkspaceClient
-from databricks.sdk.service.sql import Query
+from databricks.sdk.service.sql import Query, Dashboard
+from databricks.sdk.errors.platform import NotFound
 
 from databricks.labs.ucx.source_code.base import Fixer
 
+logger = logging.getLogger(__name__)
+
 
 class Redash:
+    MIGRATED_TAG = "migrated by UCX"
+    BACKUP_TAG = "backup by UCX"
+
     def __init__(self, fixer: Fixer, ws: WorkspaceClient):
         self._fixer = fixer
         self._ws = ws
 
-    def fix(self, query: Query):
+    def fix_dashboard(self, dashboard_id: str | None = None):
+        if dashboard_id is None:
+            dashboards = list(self._ws.dashboards.list())
+        else:
+            dashboards = [Dashboard.from_dict(self._get_dashboard(dashboard_id))]
+        for dashboard in dashboards:
+            if dashboard.tags is not None and self.MIGRATED_TAG in dashboard.tags:
+                # already migrated
+                return
+            for query in self._get_queries_from_dashboard(dashboard):
+                self._fix_query(query)
+
+    def revert_dashboard(self, dashboard_id: str | None = None):
+        if dashboard_id is None:
+            dashboards = list(self._ws.dashboards.list())
+        else:
+            dashboards = [Dashboard.from_dict(self._get_dashboard(dashboard_id))]
+        for dashboard in dashboards:
+            if dashboard.tags is None or self.MIGRATED_TAG not in dashboard.tags:
+                logger.debug(f"Dashboard {dashboard.id} was not migrated by UCX")
+                return
+            for query in self._get_queries_from_dashboard(dashboard):
+                self._revert_query(query)
+
+    def delete_backup_dbsql_queries(self):
+        for query in self._ws.queries.list():
+            if query.tags is None or self.BACKUP_TAG not in query.tags:
+                continue
+            try:
+                self._ws.queries.delete(query.id)
+            except NotFound:
+                continue
+
+    def _fix_query(self, query: Query):
         assert query.id is not None
         assert query.query is not None
-        query.query = self._fixer.apply(query.query)
-        self._ws.queries.update(
-            query.id,
+        if query.tags is not None and self.MIGRATED_TAG in query.tags:
+            # already migrated
+            return
+
+        backup_query = self._ws.queries.create(
```

Comment:
create a folder if it doesn't exist so that all queries are in the `self._installation.install_folder() / f'backup/queries')`